File: pyfile/fullsample_gmm_dml/tests/test_covariance.py
from __future__ import annotations

from itertools import combinations

import numpy as np
import pandas as pd
import pytest

from fullsample_gmm import compute_cluster_moment_matrix


CLUSTERS = pd.DataFrame(
    {
        "one": [0, 0, 0, 1, 1, 2, 2, 2, 3, 3, 3, 3],
        "two": [0, 0, 1, 0, 1, 0, 1, 2, 0, 1, 2, 2],
        "three": ["a", "b", "a", "b", "c", "a", "c", "b", "a", "c", "b", "c"],
        "four": [True, False, True, True, False, False, True, False, True, False, True, False],
    }
)
SCORES = np.random.default_rng(2026).normal(size=(len(CLUSTERS), 3))


def brute_force(scores, data, columns, method):
    labels = data.loc[:, columns].to_numpy()
    q = scores.shape[1]
    total = np.zeros((q, q))
    if method == "PSD":
        subsets = [((index,), 1.0) for index in range(len(columns))]
    else:
        subsets = []
        for size in range(1, len(columns) + 1):
            sign = 1.0 if size % 2 else -1.0
            subsets.extend(
                (subset, sign)
                for subset in combinations(range(len(columns)), size)
            )

    for subset, sign in subsets:
        for left in range(len(data)):
            for right in range(len(data)):
                if all(labels[left, level] == labels[right, level] for level in subset):
                    total += sign * np.outer(scores[left], scores[right])
    counts = [data[column].nunique() for column in columns]
    return min(counts) / len(data) ** 2 * total


@pytest.mark.parametrize("n_dimensions", [1, 2, 3, 4])
@pytest.mark.parametrize("method", ["PSD", "CGM"])
def test_k_way_matrix_matches_pairwise_definition(n_dimensions, method):
    columns = list(CLUSTERS.columns[:n_dimensions])
    result = compute_cluster_moment_matrix(SCORES, CLUSTERS, columns, method)
    expected = brute_force(SCORES, CLUSTERS, columns, method)

    np.testing.assert_allclose(result.matrix, expected, atol=1e-12)
    assert result.n_eff == min(CLUSTERS[column].nunique() for column in columns)
    assert result.n_cells == len(CLUSTERS.groupby(columns))


@pytest.mark.parametrize("n_dimensions", [1, 2, 3, 4])
def test_psd_is_positive_semidefinite(n_dimensions):
    columns = list(CLUSTERS.columns[:n_dimensions])
    matrix = compute_cluster_moment_matrix(
        SCORES, CLUSTERS, columns, "PSD"
    ).matrix
    assert np.linalg.eigvalsh(matrix).min() >= -1e-12


def test_permutation_and_cluster_relabeling_do_not_change_matrix():
    columns = list(CLUSTERS.columns)
    original = compute_cluster_moment_matrix(
        SCORES, CLUSTERS, columns, "PSD"
    ).matrix
    order = np.random.default_rng(4).permutation(len(CLUSTERS))
    permuted_data = CLUSTERS.iloc[order].reset_index(drop=True).copy()
    permuted_scores = SCORES[order]
    permuted_data["one"] = permuted_data["one"].map(
        {0: "north", 1: "south", 2: "east", 3: "west"}
    )
    changed = compute_cluster_moment_matrix(
        permuted_scores, permuted_data, columns, "PSD"
    ).matrix
    np.testing.assert_allclose(changed, original, atol=1e-12)


@pytest.mark.parametrize(
    "change, message",
    [
        (lambda frame: frame.assign(one=np.nan), "missing"),
        (lambda frame: frame.assign(one=1), "at least two"),
    ],
)
def test_invalid_cluster_identifiers_are_rejected(change, message):
    with pytest.raises(ValueError, match=message):
        compute_cluster_moment_matrix(SCORES, change(CLUSTERS), ["one"], "PSD")


def test_invalid_cluster_columns_and_method_are_rejected():
    with pytest.raises(ValueError, match="repeated"):
        compute_cluster_moment_matrix(SCORES, CLUSTERS, ["one", "one"], "PSD")
    with pytest.raises(ValueError, match="not found"):
        compute_cluster_moment_matrix(SCORES, CLUSTERS, ["absent"], "PSD")
    with pytest.raises(ValueError, match="exactly"):
        compute_cluster_moment_matrix(SCORES, CLUSTERS, ["one"], "psd")


def test_invalid_scores_are_rejected():
    with pytest.raises(ValueError, match="two-dimensional"):
        compute_cluster_moment_matrix(SCORES[:, 0], CLUSTERS, ["one"], "PSD")
    invalid = SCORES.copy()
    invalid[0, 0] = np.nan
    with pytest.raises(ValueError, match="finite"):
        compute_cluster_moment_matrix(invalid, CLUSTERS, ["one"], "PSD")

File: fullsampleDML-r/tests/testthat/test-covariance.R
test_that("cluster matrices match cross-language golden values", {
  data <- data.frame(
    one = c(0, 0, 1, 1),
    two = c(0, 1, 0, 1)
  )
  scores <- matrix(c(1, 2, 3, 4, 5, 6, 7, 8), byrow = TRUE, ncol = 2)
  fixture <- utils::read.csv(test_path("fixtures", "cluster_moments.csv"))
  for (method in c("PSD", "CGM")) {
    expected_rows <- fixture[fixture$method == method, ]
    expected <- matrix(expected_rows$value, 2, 2, byrow = TRUE)
    result <- compute_cluster_moment_matrix(scores, data, c("one", "two"), method)
    expect_equal(result$matrix, expected, tolerance = 1e-12)
    expect_equal(result$n_eff, 2L)
    expect_equal(result$n_cells, 4L)
  }
})

brute_force_middle <- function(scores, data, columns, method) {
  q <- ncol(scores)
  total <- matrix(0, q, q)
  subsets <- unlist(lapply(seq_along(columns), function(size) {
    utils::combn(seq_along(columns), size, simplify = FALSE)
  }), recursive = FALSE)
  if (method == "PSD") subsets <- lapply(seq_along(columns), identity)
  for (subset in subsets) {
    sign <- if (method == "PSD" || length(subset) %% 2L) 1 else -1
    for (left in seq_len(nrow(data))) {
      for (right in seq_len(nrow(data))) {
        same <- vapply(subset, function(index) {
          data[[columns[index]]][left] == data[[columns[index]]][right]
        }, logical(1))
        if (all(same)) total <- total + sign * tcrossprod(scores[left, ], scores[right, ])
      }
    }
  }
  counts <- vapply(data[columns], function(x) length(unique(x)), integer(1))
  min(counts) / nrow(data)^2 * total
}

test_that("one- through four-way matrices match the pairwise definition", {
  clusters <- data.frame(
    one = c(0, 0, 0, 1, 1, 2, 2, 2, 3, 3, 3, 3),
    two = c(0, 0, 1, 0, 1, 0, 1, 2, 0, 1, 2, 2),
    three = c("a", "b", "a", "b", "c", "a", "c", "b", "a", "c", "b", "c"),
    four = c(TRUE, FALSE, TRUE, TRUE, FALSE, FALSE, TRUE, FALSE, TRUE, FALSE, TRUE, FALSE)
  )
  set.seed(2026)
  scores <- matrix(rnorm(nrow(clusters) * 3L), ncol = 3L)
  for (k in 1:4) {
    columns <- names(clusters)[seq_len(k)]
    for (method in c("PSD", "CGM")) {
      actual <- compute_cluster_moment_matrix(scores, clusters, columns, method)$matrix
      expected <- brute_force_middle(scores, clusters, columns, method)
      expect_equal(actual, expected, tolerance = 1e-12)
    }
    psd <- compute_cluster_moment_matrix(scores, clusters, columns, "PSD")$matrix
    expect_gte(min(eigen(psd, symmetric = TRUE, only.values = TRUE)$values), -1e-12)
  }
})

test_that("cluster matrices are invariant to order and relabeling", {
  data <- data.frame(a = rep(1:3, each = 4), b = rep(letters[1:4], 3))
  set.seed(4)
  scores <- matrix(rnorm(24), ncol = 2)
  original <- compute_cluster_moment_matrix(scores, data, c("a", "b"))$matrix
  order <- sample(seq_len(nrow(data)))
  changed <- data[order, ]
  changed$a <- c("north", "south", "east")[changed$a]
  relabeled <- compute_cluster_moment_matrix(scores[order, ], changed, c("a", "b"))$matrix
  expect_equal(relabeled, original, tolerance = 1e-12)
})

test_that("invalid clustered covariance inputs are rejected", {
  data <- data.frame(a = c(1, 1, 2, 2), b = c(1, 2, 1, 2))
  scores <- matrix(1:8, ncol = 2)
  expect_error(compute_cluster_moment_matrix(scores, data, c("a", "a")), "repeated")
  expect_error(compute_cluster_moment_matrix(scores, data, "absent"), "not found")
  expect_error(compute_cluster_moment_matrix(scores, data, "a", "psd"), "exactly")
  expect_error(compute_cluster_moment_matrix(scores[, 1], data, "a"), "two-dimensional")
  bad <- scores; bad[1, 1] <- NA_real_
  expect_error(compute_cluster_moment_matrix(bad, data, "a"), "finite")
  expect_error(compute_cluster_moment_matrix(scores, transform(data, a = NA), "a"), "missing")
  expect_error(compute_cluster_moment_matrix(scores, transform(data, a = 1), "a"), "at least two")
})
